From f0928bc8c9281894eee38987a90fcd8fdc882d74 Mon Sep 17 00:00:00 2001
From: Chetan Nanda <chetan.nanda@st.com>
Date: Thu, 17 Sep 2015 16:19:55 +0530
Subject: [PATCH 15/15] v4l-utils: libv4l-hva: new ops 'set_ctrl' added

With dynamic ctrls support at v4l2enc, all ctrls may not be set
at same time.
So, ctrls values should be updated to encoder header generation code
whenever any value change occur.

Change-Id: Ib168e60751df6eb6177d18376f81d9d5c9807017
Signed-off-by: Chetan Nanda <chetan.nanda@st.com>
Reviewed-on: https://gerrit.st.com/38429
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Reviewed-by: Yannick FERTRE <yannick.fertre@st.com>
---
 lib/libv4l-hva/libv4l-hva-h264.c | 4 +---
 lib/libv4l-hva/libv4l-hva.c      | 2 ++
 lib/libv4l-hva/libv4l-hva.h      | 7 +++++++
 3 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/lib/libv4l-hva/libv4l-hva-h264.c b/lib/libv4l-hva/libv4l-hva-h264.c
index 5aaf221..dc1bb05 100644
--- a/lib/libv4l-hva/libv4l-hva-h264.c
+++ b/lib/libv4l-hva/libv4l-hva-h264.c
@@ -237,9 +237,6 @@ hva_h264_get_sps_pps(void *priv, uint8_t **data, size_t *size,
 	else
 		return -EINVAL;
 
-	/* update controls */
-	hva_h264_set_ctrl(priv, ctrls);
-
 	memset (&enc_h264, 0, sizeof(enc_h264));
 	memset (&picture, 0, sizeof(picture));
 	memset (&sequence, 0, sizeof(sequence));
@@ -402,5 +399,6 @@ const struct hva_metadata h264meta = {
 	.start = hva_h264_start,
 	.stop = hva_h264_stop,
 	.get_sps_pps = hva_h264_get_sps_pps,
+	.set_ctrl = hva_h264_set_ctrl,
 };
 
diff --git a/lib/libv4l-hva/libv4l-hva.c b/lib/libv4l-hva/libv4l-hva.c
index 908dbcc..11a8610 100644
--- a/lib/libv4l-hva/libv4l-hva.c
+++ b/lib/libv4l-hva/libv4l-hva.c
@@ -137,6 +137,8 @@ static int hva_set_ext_ctrls(struct hva_plugin *hva, int fd,
 
 	hva->ctrls.count = ctrls->count;
 
+	/* update ctrls  */
+	hva->meta->set_ctrl(hva->meta_priv, &hva->ctrls);
 	ret = SYS_IOCTL(fd, cmd, ctrls);
 	if (ret) {
 		V4L2_LOG_ERR("%s: failed to set controls\n", __func__);
diff --git a/lib/libv4l-hva/libv4l-hva.h b/lib/libv4l-hva/libv4l-hva.h
index 338bbab..5564fb9 100644
--- a/lib/libv4l-hva/libv4l-hva.h
+++ b/lib/libv4l-hva/libv4l-hva.h
@@ -72,6 +72,13 @@ struct hva_metadata {
 	int (*get_sps_pps) (void *priv, uint8_t **data, size_t *size,
 						struct v4l2_format *format,
 						struct v4l2_ext_controls *ctrls);
+
+	/**
+	 * set_ctrl() - update the external ctrl values to header encoder
+	 * @ctrls: (in) external controls
+	 *
+	 */
+	int (*set_ctrl) (void *priv, struct v4l2_ext_controls *ctrls);
 };
 
 /**
-- 
2.7.4

